File: src/Montague/Display.hs

module Montague.Display where

import Montague.Types

-- | Helper function to display a parsed term as a prolog term.
displayAsPrologTerm :: Show a => Term a -> String
displayAsPrologTerm x = show x

-- | Helper function to display a parsed term as an s-expression.
displayAsSExp :: Show a => Term a -> String
displayAsSExp (Atom xs) = show xs
displayAsSExp (App x xs) = "(" ++ show x ++ (foldr1 (\x -> \y -> x ++ "," ++ y) (map show xs)) ++ ")"

File: montague-core/src/display.rs
use crate::types::Term;

/// Format a term as a Prolog term: `f(x,y)`.
pub fn display_as_prolog_term<A: std::fmt::Display>(term: &Term<A>) -> String {
    term.to_string()
}

/// Format a term as an S-expression: `(f x y)`.
pub fn display_as_sexp<A: std::fmt::Display>(term: &Term<A>) -> String {
    match term {
        Term::Atom(a) => a.to_string(),
        Term::Var(s) => s.clone(),
        Term::Lambda(s, body) => format!("(λ {} {})", s, display_as_sexp(body)),
        Term::App(f, args) => {
            let mut parts = vec![display_as_sexp(f)];
            parts.extend(args.iter().map(display_as_sexp));
            format!("({})", parts.join(" "))
        }
    }
}
